Tests/SonarScanner.MSBuild.PreProcessor.Test/WebClientDownloaderTest.cs

```diff
@@ -146,26 +150,156 @@ public async Task Download_Fail_NotForbidden_WithLog()
+
+            result.Should().BeFalse();
+            logger.AssertDebugLogged($"Downloading file from {BaseUrl}{RelativeUrl} to {fileName}...");
             logger.AssertNoWarningsLogged();
         }
 
         [TestMethod]
-        [DataRow("https://sonarsource.com/", "https://sonarsource.com/")]
-        [DataRow("https://sonarsource.com", "https://sonarsource.com/")]
-        [DataRow("https://sonarsource.com/sonarlint", "https://sonarsource.com/sonarlint/")]
-        [DataRow("https://sonarsource.com/sonarlint/", "https://sonarsource.com/sonarlint/")]
-        public void GetBaseUri_ValidUrl_ShouldAlwaysEndsWithSlash(string baseUrl, string expectedUrl)
+        public async Task TryDownloadFileIfExists_HttpCodeForbidden_ShouldThrowWithLog()
         {
-            var sut = new WebClientDownloader(new HttpClient(), baseUrl, Mock.Of<ILogger>());
+            var logger = new TestLogger();
+            var sut = CreateSut(logger, HttpStatusCode.Forbidden);
+            var fileName = Path.GetRandomFileName();
 
-            var result = sut.GetBaseUri();
+            Func<Task> act = async () => await sut.TryDownloadFileIfExists(RelativeUrl, fileName);
+
+            await act.Should().ThrowExactlyAsync<HttpRequestException>();
+            logger.AssertDebugLogged($"Downloading file from {BaseUrl}{RelativeUrl} to {fileName}...");
+            logger.AssertNoWarningsLogged();
+        }
+
+        [TestMethod]
+        public async Task TryDownloadFileIfExists_HttpCodeForbiddenWithWarnLog_ShouldThrowWithLog()
+        {
+            var logger = new TestLogger();
+            var sut = CreateSut(logger, HttpStatusCode.Forbidden);
+            var fileName = Path.GetRandomFileName();
 
-            result.ToString().Should().EndWith("/");
-            result.ToString().Should().Be(expectedUrl);
+            Func<Task> act = async () => await sut.TryDownloadFileIfExists(RelativeUrl, fileName, true);
+
+            await act.Should().ThrowExactlyAsync<HttpRequestException>();
+            logger.AssertDebugLogged($"Downloading file from {BaseUrl}{RelativeUrl} to {fileName}...");
+            logger.AssertWarningLogged("To analyze private projects make sure the scanner user has 'Browse' permission.");
+        }
+
+        [TestMethod]
+        [DynamicData(nameof(UnsuccessfulHttpCodeData))]
+        public async Task TryDownloadFileIfExists_UnsuccessfulHttpCode_ShouldThrowWithLog(HttpStatusCode code)
+        {
+            var logger = new TestLogger();
+            var sut = CreateSut(logger, code);
+            var fileName = Path.GetRandomFileName();
+
+            Func<Task> act = async () => await sut.TryDownloadFileIfExists(RelativeUrl, fileName);
+
+            await act.Should().ThrowExactlyAsync<HttpRequestException>();
+            logger.AssertDebugLogged($"Downloading file from {BaseUrl}{RelativeUrl} to {fileName}...");
+            logger.AssertNoWarningsLogged();
```

Comment:
Maybe I've missed it but I could not find a test that check the url concatenation. I would like to have a test with base url "https://www.sonarsource.com/" and one with "https://www.sonarsource.com/virtualdir" and see the resulting url after concatenation with the rest of the parts.